geomesa-raster/src/main/scala/org/locationtech/geomesa/raster/util/RasterUtils.scala

```diff
@@ -83,46 +107,110 @@ object RasterUtils {
+    val properties = new JHashtable[String, Object]
+    if (r.getPropertyNames != null) {
+      r.getPropertyNames.foreach(name => properties.put(name, r.getProperty(name)))
+    }
+    val colorModel = r.getColorModel
+    val alphaPremultiplied = colorModel.isAlphaPremultiplied
+    val sampleModel = r.getSampleModel
+    new BufferedImage(colorModel, r.copyData(null), alphaPremultiplied, properties)
   }
 
-  def getEmptyImage(width: Int, height: Int): BufferedImage = {
-    new BufferedImage(width, height, BufferedImage.TYPE_BYTE_GRAY)
+  def getEmptyImage(width: Int, height: Int, imageType: Int): BufferedImage = {
+    new BufferedImage(width, height, imageType)
   }
 
-  def mosaicRasters(rasters: Iterator[Raster], width: Int, height: Int, env: Envelope, resX: Double, resY: Double): (BufferedImage, Int) = {
-    if (rasters.isEmpty) {
-      (getEmptyImage(width, height), 0)
+  def writeToMosaic(mosaic: BufferedImage, raster: Raster, env: Envelope, resX: Double, resY: Double) = {
+    val croppedRaster = cropRaster(raster, env)
+    croppedRaster.map { cropped =>
+      val rasterEnv = raster.referencedEnvelope.intersection(envelopeToReferencedEnvelope(env))
+      val originX = Math.floor((rasterEnv.getMinX - env.getMinimum(0)) / resX).toInt
+      val originY = Math.floor((env.getMaximum(1) - rasterEnv.getMaxY) / resY).toInt
+      mosaic.getRaster.setRect(originX, originY, cropped.getData)
+      cropped.flush()
+    }
+  }
+
+  def mosaicChunks(chunks: Iterator[Raster], queryWidth: Int, queryHeight: Int, queryEnv: Envelope): (BufferedImage, Int) = {
+    if (chunks.isEmpty) {
+      (getEmptyImage(queryWidth, queryHeight, BufferedImage.TYPE_BYTE_GRAY), 0)
     } else {
-      val rescaleX = resX / (env.getSpan(0) / width)
-      val rescaleY = resY / (env.getSpan(1) / height)
-      val scaledWidth = width / rescaleX
-      val scaledHeight = height / rescaleY
-      val imageWidth = Math.max(Math.round(scaledWidth), 1).toInt
-      val imageHeight = Math.max(Math.round(scaledHeight), 1).toInt
-      val firstRaster = rasters.next()
       var count = 1
-      val mosaic = getEmptyMosaic(imageWidth, imageHeight, firstRaster.chunk)
-      setMosaicData(mosaic, firstRaster, env, resX, resY)
-      while (rasters.hasNext) {
-        val raster = rasters.next()
-        setMosaicData(mosaic, raster, env, resX, resY)
-        count += 1
+      val firstRaster = chunks.next()
+      val accumuloRasterXRes = firstRaster.referencedEnvelope.getSpan(0) / firstRaster.chunk.getWidth
+      val accumuloRasterYRes = firstRaster.referencedEnvelope.getSpan(1) / firstRaster.chunk.getHeight
+      val mosaicX = (queryEnv.getSpan(0) / accumuloRasterXRes).toInt
+      val mosaicY = (queryEnv.getSpan(1) / accumuloRasterYRes).toInt
+      if (mosaicX <= 0 || mosaicY <= 0) {
+        (getEmptyImage(1, 1, BufferedImage.TYPE_BYTE_GRAY), count)
+      } else {
+        val mosaic = allocateBufferedImage(mosaicX, mosaicY, firstRaster.chunk)
```

Comment:
I think we currently handle this case in a reasonable way, we always mosaic the first image in the iterator and then only do further work if the iterator hasNext.